app/models/policy.rb

```diff
@@ -50,12 +50,25 @@ def update_hosts(new_host_ids)
     return unless new_host_ids
 
     removed = policy_hosts.where.not(host_id: new_host_ids).destroy_all
-    imported = PolicyHost.import((new_host_ids - host_ids).map do |host_id|
-      { host_id: host_id, policy_id: id }
-    end)
+    imported = PolicyHost.import_from_policy(id, new_host_ids - host_ids)
+    update_os_versions
+
     [imported.ids.count, removed.count]
   end
 
+  def update_os_versions
+    os_versions = Host.os_versions(hosts.pluck(:id))
+    initial_profile.update(os_version: os_versions.pop) if os_versions.any?
```

Comment:
Setting of the OS info on the initial profile seems non-deterministic. Initial profile should get the OS info set only if it is from the benchmark that an OS has a latest support. Once set, the OS info must not be changed.